std/experimental/color/rgb.d

```
@@ -0,0 +1,852 @@
+
+    // create members for some useful information
+    alias ComponentType = ComponentType_;
+    enum string components = components_;
+    enum RGBColorSpace colorSpace = colorSpace_;
+    enum bool linear = linear_;
+    enum bool hasAlpha = CanFind!(components, 'a');
+
+
+    // mixin will emit members for components
+    template Components(string components)
+    {
+        static if(components.length == 0)
+            enum Components = "";
+        else
+            enum Components = (components[0] == 'x' ? "private " : "") ~ ComponentType.stringof ~ ' ' ~ components[0] ~ " = 0;\n" ~ Components!(components[1..$]);
+    }
+    mixin(Components!components);
+
+    // missing components will be fabricated such that code that references them compiles
+    mixin template MissingComponent(char name)
+    {
+        // missing color components will equal the luminance, if it exists. 0 otherwise.
+        static if(!CanFind!(components, 'l'))
+            mixin("enum ComponentType "~name~" = 0;");
+        else
+            mixin("alias "~name~" = l;");
+    }
+    static if(!CanFind!(components, 'r'))
+        mixin MissingComponent!'r';
+    static if(!CanFind!(components, 'g'))
+        mixin MissingComponent!'g';
+    static if(!CanFind!(components, 'b'))
+        mixin MissingComponent!'b';
+    static if(!CanFind!(components, 'l'))
+        @property ComponentType l() const { return toGrayscale!colorSpace(r, g, b); }
+    static if(!CanFind!(components, 'a'))
+        enum ComponentType a = 0;
+
+    // functions that return the color channels as a tuple
+    static if(CanFind!(components, 'l'))
+        @property auto tristimulus() const { auto t = l; return tuple(t, t, t); }
+    else
+        @property auto tristimulus() const { return tuple(r, g, b); }
+    @property auto tristimulusWithAlpha() const { return tuple(tristimulus.expand, a); }
+
+    // RGB/A initialiser
+    this(ComponentType r, ComponentType g, ComponentType b, ComponentType a = 0)
```

I debated this with myself for ages. Ultimately, I concluded that a default value needs to be 'identity' otherwise various constructs start failing. 0 is the scalar identity.
Trouble with a=1, is that when you add 2 colours, a==2, and that's no good.
Since it's a silent default arg, and it may be used in arithmetic, I think the risk of it perturbing users results is high without their noticing. Conversely, they will certainly notice immediately if they forgot to specify a=1 when constructing a colour with an alpha channel.